### PR TITLE
Escaping spawn

### PR DESCRIPTION
**Pull request by lucteo**
Add support for `spawn` in which the `await` is made in a different scope.
In this case, the frame object needs to be allocated on the heap.

### CODE REVIEW

File: StandardLibrary/Sources/Concurrency/Future.hylo [hunk 1]
```diff
@@ -14,35 +68,32 @@ public type Future<E: Movable & Deinitializable> {
   private var r: Optional<Int>
 
   /// Initializes `self` with `f`, and spawns the computation.
-  public init(_ f: sink [E]() -> Int) {
+  public fun manual_init(_ f: sink [E]() -> Int) set {
```
Comment (kyouko-taiga) [inline StandardLibrary/Sources/Concurrency/Future.hylo:71]: Why did you call it `manual_init`. You can call `init` "manually" as a static method if you want:

```
type A: Deinitializable {
  public init(x: Int) {}
}

public fun main() {
  var a: A
  A.`init`(self: &a, x: 1)
}
```

File: StandardLibrary/Sources/Concurrency/Future.hylo [hunk 2]
```diff
@@ -4,6 +4,60 @@ public type Future<E: Movable & Deinitializable> {
...
+
+  // TODO
+  public typealias Result = Int
+
+  /// The underlying frame object that is used to connect the spawned computation with the await point.
+  private var frame: PointerToMutable<SpawnFrame<E>>
+
+  /// Initializes `self` with `f`, and spawns the computation.
+  public init(_ f: sink [E]() -> Int) {
+    &self.frame = alloc_frame(f)
```
Comment (kyouko-taiga) [inline StandardLibrary/Sources/Concurrency/Future.hylo:35]:
```suggestion
    &self.frame = allocate_frame(f)
```

File: StandardLibrary/Sources/Concurrency/Future.hylo [hunk 3]
```diff
@@ -4,6 +4,60 @@ public type Future<E: Movable & Deinitializable> {
...
+  /// Awaits the result of the computation.
+  ///
+  /// - Note: May return on a different OS thread than the one that called this.
+  public fun await() sink -> Result {
+    let r = (&frame.unsafe[]).await()
+    frame.deallocate()
+    return r
+  }
+
+  private static fun alloc_frame(_ f: sink [E]() -> Int) -> PointerToMutable<SpawnFrame<E>> {
```
Comment (kyouko-taiga) [inline StandardLibrary/Sources/Concurrency/Future.hylo:47]: Let's try to document every function we write in the standard library, even if they are private.

**Review by kyouko-taiga — CHANGES_REQUESTED** [review]
Nice PR, thanks. I'm not convinced that defining an initializer with a custom name is justified, but otherwise it looks good. My other comments are mostly about style.

File: StandardLibrary/Sources/Concurrency/Future.hylo (hunk 1 above)
Comment (lucteo) [reply StandardLibrary/Sources/Concurrency/Future.hylo]: Fixed. Thank you, I did not know that.

File: StandardLibrary/Sources/Concurrency/Future.hylo (hunk 2 above)
Comment (lucteo) [reply StandardLibrary/Sources/Concurrency/Future.hylo]: Moved it as internal method of `SpawnFrame`.

File: StandardLibrary/Sources/Concurrency/Future.hylo (hunk 3 above)
Comment (lucteo) [reply StandardLibrary/Sources/Concurrency/Future.hylo]: Fixed. Thank you.

**Review by kyouko-taiga — APPROVED** [review]
LGTM

**Pull request merged**